### ISSUE TITLE
Intermittent (but frequent) failures to install packages, since 9.5.0

### ISSUE THREAD

**Issue by lanthonyneville** [dnnsoftware/Dnn.Platform#3681]
Since installing 9.5.0, I have experienced frequent failures while installing module packages. The failures have been seen on 3 independent DNN installations. The failures were never seen before 9.5.0.

The failures are not reliably reproducible. The system seems to get into a state where every attempt to install fails for a stretch of time, then after waiting a few minutes, the same install will work.

I have seen this bug both when upgrading an existing module to a new version, and when installing a completely new module. I have not tried with non-module packages.

### Description of bug
1. Log in as a super user
2. Navigate to Settings > Extensions (in Persona Bar)
3. Click Install Extension
4. Drag and drop a module install package onto the target area.
5. At this point there are 2 possible ways the failure manifests:

## Failure Type 1
- The drag and drop target area immediately displays the error: Zip File Upload Failed
- On clicking View Error Log, this message appears:
![image](https://user-images.githubusercontent.com/22200046/79333291-30071500-7f16-11ea-830f-c1ec0b58eab5.png)
- On clicking Try Again, the error may repeat, or the process may continue to the next step and fail with Failure Type 2, or the install may complete successfully

## Failure Type 2
- The process continues until the License page
- On checking the Accept License box and clicking Next, the Next button disables, but no progress is seen. The UI "hangs", although it is possible to click Cancel and return to the beginning.
- On clicking Cancel, the error may repeat, or the install may complete successfully

### Error seen in Event Log
The following entry is seen for every failure:

AbsoluteURL:/API/PersonaBar/Extensions/InstallPackage
DefaultDataProvider:DotNetNuke.Data.SqlDataProvider, DotNetNuke
ExceptionGUID:f2cae52e-621c-4a30-aa35-189920a54651
AssemblyVersion:
PortalId:-1
UserId:-1
TabId:-1
RawUrl:
Referrer:
UserAgent:
ExceptionHash:Y7Roa4W/K5qsKcQ/vfVCIltK4UQ=
Message:The operation was canceled.
StackTrace:
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__0.MoveNext()

InnerMessage:
InnerStackTrace:

Source:mscorlib
FileName:
FileLineNumber:0
FileColumnNumber:0
Method:

**Comment by mitchelsellers** [comment]
Are there any corresponding errors in the /portals/_default/logs folder at the time

**Comment by lanthonyneville** [comment]
Here is a sample log file from the /portals/_default/logs folder .... it seems to show the same error seen in the Event Log.
[2020.04.15.log.txt](https://github.com/dnnsoftware/Dnn.Platform/files/4481628/2020.04.15.log.txt)

**Comment by EPTamminga** [comment]
I had the same experience. 
My solution was to wait and refresh the home page of the site after each individual extension upload, so no quick install of 2 packages directly after each other.
Upload an extension pack, wait until it seems that the extension is processed, go to your home page, refresh, wait until that has finished, than continue with the next extension install package.

**Comment by valadas** [comment]
Could this be with specific modules, can you guys give some examples of modules you installed maybe to see if there is something in common. I have not experienced this issue.

**Comment by EPTamminga** [comment]
I experienced this while updating the DNNSharp modules (Grid, Form) in DNN932 as part of an intermediate step to the final update to DNN950

**Comment by lanthonyneville** [comment]
I had had this issue installing [OpenContent 4.4](https://github.com/sachatrauwaen/OpenContent) and several modules developed by myself (sorry - can't share as private to company).

**Comment by lanthonyneville** [comment]
Just noticed something interesting .....

In an attempt to find out which request was failing, I ran Fiddler and attempted a module upgrade (a re-install of OpenContent 4.4). It worked. I then re-applied the upgrade several times in rapid succession, and every one worked.

I then stopped Fiddler, and re-applied the upgrade. It immediately failed with Failure Type 1. I tried again and it failed again. I then restarted Fiddler, and it immediately worked.

Fiddler acts as a proxy, so all requests go through it before going to IIS. I have no idea however why a request going through Fiddler would not be cancelled!

**Comment by valadas** [comment]
>sorry - can't share as private to company

Cool, understandable, was just looking at trying to find if the people experiencing this where experiencing it with the same modules or what.. Intermittent issues are a pain :) Until we find out why it happens, there is no way to fix it...

The fiddler thing is very odd indeed. Just as a test was this tried using different machines, different browswers, incognito mode or some such ?

**Comment by lanthonyneville** [comment]
I have tried with Chrome in normal mode and incognito, Edge and IE 11, and experienced the issue on all of them. 

From doing some more testing, it seems that the issue is related to ZIP file size. My smallest package (113KB) installs reliably, but my biggest (8MB) regularly fails to install.

However it isn't as simple as enabling large file uploads in web.config. I have the following under `<system.webServer>` already, and like I said installs sometimes work and sometimes don't and always work when Fiddler is running .....

```
    <security>
      <requestFiltering>
        <requestLimits maxAllowedContentLength="30000000" />
      </requestFiltering>
    </security>
```

**Comment by david-poindexter** [comment]
I am wondering if this is perhaps environmental.  I have yet to experience this and have built close to 100 sites now on 9.5.0.  That said, our module set has been minimal.  However, several of the modules have very large install packages (still small enough though for the default `httpRuntime` settings).

**Comment by mitchelsellers** [comment]
Is this by chance on the Azure environment?

As certain third-party modules grow in size they have started to include up to 1,000's of installation files and that can slow progress on installation or inspection and hit limits.

Also, how quick does the error appear?  Is it instant or after a delay?

**Comment by lanthonyneville** [comment]
I guess it is possible that it is the environment, however the issue seems to coincide exactly with installing 9.5.0. I have a number of environments on which I started to get this issue as soon as I upgraded, and before that I had never seen it. Possibly there is something common in the config for these environments however ....

These are on-premises Windows 2016 and IIS 10.0 environments, not Azure.

@mitchelsellers - see the first post for how the issue shows up - if its Type 1 it is instantly after dragging the ZIP over the target panel. If it is Type 2 there is a delay as the installer appears to be processing the file but then fails to complete.

**Comment by bdukes** [comment]
Since it doesn't happen with Fiddler open, does your environment have a default proxy that's typically capturing your traffic?  Or is there a firewall or some other appliance between you and the server which could be having trouble with the traffic?

**Comment by lanthonyneville** [comment]
Ah ha! I did another experiment and it seems to prove the issue is in the network we usually use to do the installs from, and not the DNN app or the configuration of the site itself.
I did a few installs from a separate network, and they all worked.
There must have been some change in the network, or threat protection software on the workstations that is cancelling the large requests, and it must have come in around 9.5.0 being released.
Thanks for your help here. Hope this is a useful as a search result for anyone with the same issue!
